File: Recitations-F2007/Recitation-06/python/exercise10.py
#!/usr/bin/python

# 
# Working definitions (python)
#
def make_units(C,L,H):
  return [C,L,H]
def get_units_C(x):
  return x[0]
def get_units_L(x):
  return x[1]
def get_units_H(x):
  return x[2]

def make_class(number,units):
  return [number,units]
def get_class_number(x):
  return x[0]
def get_class_units(x):
  return x[1]

def get_class_total_units(klass):
  units = get_class_units(klass)
  return get_units_C(units) + get_units_L(units) + get_units_H(units)

def same_class(klass1,klass2):
  return get_class_number(klass1) == get_class_number(klass2)

#
# Working definitions (HOPs)
#
def make_student(number,sched_checker):
  return [number, [], sched_checker]

def get_student_number(student):
  return student[0]

def get_student_schedule(student):
  return student[1]

def get_student_checker(student):
  return student[2]

def update_student_schedule(student, schedule):
  checker = get_student_checker(student)
  if ( checker(schedule) ):
    return [get_student_number(student), schedule, checker]
  else:
    raise BaseException("Invalid schedule!")

#
# Previous Solutions
#
def empty_schedule():
  return []

def add_class(klass,schedule):
  schedule.append(klass)
  return schedule

def total_scheduled_units(schedule):
  def total_scheduled_units_iter(total,working):
    if not working:
      return total
    else:
      return total_scheduled_units_iter(total + get_class_total_units(working[0]), working[1:])
  return total_scheduled_units_iter(0,schedule)

def drop_class(schedule,classnum):
  tmp = make_class(classnum,[])
  def predicate(klass):
    return not same_class(klass,tmp)
  return filter(predicate, schedule)

def credit_limit(schedule, max_credits):
  def credit_limit_iter(working):
    if not working:
      return []
    else:
      total_credits = total_scheduled_units(working)
      first_class = working[0]
      if ( total_credits > max_credits ):
        return credit_limit_iter(drop_class(working, get_class_number(first_class)))
      else:
        return working
  return credit_limit_iter(schedule)

def make_schedule_checker_1():
  return lambda x: len(x) > 0

def make_schedule_checker_2(max_credits):
  return lambda x: total_scheduled_units(x) <= max_units

def class_numbers(schedule):
  return map(lambda x: get_class_number(x), schedule)

# 
# Basic Classes
#
calc1 = make_class("CALC-101", make_units(4,4,4))
calc2 = make_class("CALC-102", make_units(4,4,4))
algebra = make_class("ALGB-152", make_units(3,3,3))
diff_eqs = make_class("DIFF-201", make_units(3,3,3))
us_history = make_class("HIST-122", make_units(4,4,4))
world_history = make_class("HIST-324", make_units(4,4,4))
basket_weaving = make_class("BASKETS", make_units(1,1,1))

#
# Exercise 10
#
# Rewrite "credit-limit" to run in O(n) time.
#
def credit_limit(schedule, max_credits):
 def credit_limit_iter(sched, working, total):
  if len(sched) == 0:
   return working
  else:
   klass = sched[0]
   credits = get_class_total_units(klass)
   if (credits+total) > max_credits:
    return working
   else:
     if working is None: working = []
     working = working.append(klass)
     return credit_limit_iter(sched[1:], working, (credits+total))
 return credit_limit_iter(schedule,[],0)

#
# Let's run some unit tests:
#
s1 = empty_schedule()
s2 = empty_schedule()
s2 = add_class(calc1, s2)
s2 = add_class(algebra, s2)
s2 = add_class(diff_eqs, s2)

#
# Introspect s2:
#
s2 
# ==> [['CALC-101', [4, 4, 4]], ['ALGB-152', [3, 3, 3]], ['DIFF-201', [3, 3, 3]]]
drop_class(s2,"CALC-101")
# ==> [['ALGB-152', [3, 3, 3]], ['DIFF-201', [3, 3, 3]]]

File: Recitations-F2007/Recitation-06/clojure/exercise-10.clj
;;
;; Working definition (Lisp-style)
;;
(ns sicp.clojure.lisp)

(defn make-units [C L H] (list C L H))
(defn get-units-C [x] (first x))
(defn get-units-L [x] (first (rest x)))
(defn get-units-H [x] (first (rest (rest x))))

(defn make-class [number units] (list number units))
(defn get-class-number [x] (first x))
(defn get-class-units [x] (first (rest x)))

(defn get-class-total-units [class]
  (let [units (get-class-units class)]
    (+ (get-units-C units)
       (get-units-L units)
       (get-units-H units))))

(defn same-class? [c1 c2]
  (= (get-class-number c1) (get-class-number c2)))

;;
;; Working definitions (HOPs)
;;
(defn make-student [number sched-checker]
  (list number (list) sched-checker))
(defn get-student-number [x] (first x))
(defn get-student-schedule [x] (first (rest x)))
(defn get-student-checker [x] (first (rest (rest x))))

(defn update-student-schedule [student schedule]
  (if ((get-student-checker student) schedule)
    (list (get-student-number student)
	  schedule
	  (get-student-checker student))
    (throw (Exception. "Invalid schedule!"))))

;;
;; Previous solutions
;;
(defn empty-schedule [] '())

(defn add-class [class schedule]
  (concat schedule (list class)))

(defn total-scheduled-units [schedule]
  (defn total-scheduled-units-iter [total working]
    (if (empty? working)
      total
      (let [current-class (first working)]
	(total-scheduled-units-iter
	 (+ total
	    (get-class-total-units current-class))
	 (rest working)))))
  (total-scheduled-units-iter 0 schedule))

(defn drop-class [schedule classnum]
  (let [temp-class (make-class classnum '())]
    (defn predicate [class]
      (not (same-class? class temp-class)))
    (filter predicate schedule)))

(defn credit-limit [schedule max-credits]
  (defn credit-limit-iter [working]
    (if (empty? working)
      '()
      (let [total-credits (total-scheduled-units working)
	    first-class (first working)]
	(if (> total-credits max-credits)
	  (credit-limit-iter
	   (drop-class working (get-class-number first-class)))
	  working))))
  (credit-limit-iter schedule))

(defn make-schedule-checker-1 []
  (fn [schedule] (> (count schedule) 0)))

(defn make-schedule-checker-2 [max-units]
  (fn [schedule] (<= (total-scheduled-units schedule) max-units)))

(defn class-numbers [schedule]
  (map (fn [x] (get-class-number x)) schedule))

;;
;; Basic Classes
;;
(def calc1 (make-class 'CALC-101 (make-units 4 4 4)))
(def calc2 (make-class 'CALC-102 (make-units 4 4 4)))
(def algebra (make-class 'ALGB-152 (make-units 3 3 3)))
(def diff-eqs (make-class 'DIFF-201 (make-units 3 3 3)))
(def us-history (make-class 'HIST-122 (make-units 4 4 4)))
(def world-history (make-class 'HIST-324 (make-units 4 4 4)))
(def basket-weaving (make-class 'BASKETS (make-units 1 1 1)))

;;
;; Exercise 10
;;
;; Rewrite "credit-limit" to run in O(n) time.
;;
(defn credit-limit [schedule max-credits]
  (defn credit-limit-iter [sched working total]
    (if (empty? sched)
      working
      (let [class (first sched)]
	(let [credits (get-class-total-units class)]
	  (if (> (+ credits total) max-credits)
	    working
	    (credit-limit-iter (rest sched) (concat working (list class)) (+ credits total)))))))
  (credit-limit-iter schedule '() 0))

;;
;; Run some unit tests:
;;
(def s2 (empty-schedule))
(def s2 (add-class calculus-1 s1))
(def s2 (add-class algebra s1))
(def s2 (add-class diff-eqs s1))

(total-scheduled-units s2)
;; ==> 30

(credit-limit s2 11)
;; ==> () 
(credit-limit s2 12)
;; ==> ((CALC-101 (4 4 4)))
(credit-limit s2 20)
;; ==> ((CALC-101 (4 4 4)))
(credit-limit s2 21)
;; ==> ((CALC-101 (4 4 4)) (ALGB-152 (3 3 3)))
(credit-limit s2 29)
;; ==> ((CALC-101 (4 4 4)) (ALGB-152 (3 3 3)))
(credit-limit s2 30)
;; ==> ((CALC-101 (4 4 4)) (ALGB-152 (3 3 3)) (DIFF-201 (3 3 3)))

;; ====================================================================

;;
;; Working definitions (Joy-style)
;;
(ns sicp.clojure.joy)

(defn make-units [C L H] {:C C :L L :H H})
(defn get-units-C [x] (x :C))
(defn get-units-L [x] (x :L))
(defn get-units-H [x] (x :H))

(defn make-class [number units] {:number number :units units})
(defn get-class-number [x] (x :number))
(defn get-class-units [x] (x :units))

(defn get-class-total-units [class]
  (let [units (get-class-units class)]
    (+ (get-units-C units)
       (get-units-L units)
       (get-units-H units))))

(defn same-class? [c1 c2]
  (= (get-class-number c1) (get-class-number c2)))

;;
;; Working definitions (HOPs)
;;
(defn make-student [number sched-checker]
  {:number number :schedule [] :checker sched-checker})
(defn get-student-number [x] (x :number))
(defn get-student-schedule [x] (x :schedule))
(defn get-student-checker [x] (x :checker))

(defn update-student-schedule [student schedule]
  (if ((get-student-checker student) schedule)
    {:number (get-student-number student)
     :schedule schedule
     :checker (get-student-checker student)}
    (throw (Exception. "Invalid schedule!"))))

;;
;; Previous solutions
;;
(defn empty-schedule [] [])

(defn add-class [class schedule]
  (conj schedule class))

(defn total-scheduled-units [schedule]
  (defn total-scheduled-units-iter [total working]
    (if (empty? working)
      total
      (let [current-class (first working)]
	(total-scheduled-units-iter
	 (+ total
	    (get-class-total-units current-class))
	 (subvec working 1)))))
  (total-scheduled-units-iter 0 schedule))

(defn drop-class [schedule classnum]
  (let [temp-class (make-class classnum '())]
    (defn predicate [class]
      (not (same-class? class temp-class)))
    (vec (filter predicate schedule))))

(defn credit-limit [schedule max-credits]
  (defn credit-limit-iter [working]
    (if (empty? working)
      []
      (let [total-credits (total-scheduled-units working)
	    first-class (first working)]
	(if (> total-credits max-credits)
	  (credit-limit-iter (drop-class working (get-class-number first-class)))
	  working))))
  (credit-limit-iter schedule))

(defn make-schedule-checker-1 []
  (fn [schedule] (> (count schedule) 0)))

(defn make-schedule-checker-2 [max-units]
  (fn [schedule] (<= (total-scheduled-units schedule) max-units)))

(defn class-numbers [schedule]
  (vec (map (fn [x] (get-class-number x)) schedule)))

;;
;; Basic Classes
;;
(def calc1 (make-class 'CALC-101 (make-units 4 4 4)))
(def calc2 (make-class 'CALC-102 (make-units 4 4 4)))
(def algebra (make-class 'ALGB-152 (make-units 3 3 3)))
(def diff-eqs (make-class 'DIFF-201 (make-units 3 3 3)))
(def us-history (make-class 'HIST-122 (make-units 4 4 4)))
(def world-history (make-class 'HIST-324 (make-units 4 4 4)))
(def basket-weaving (make-class 'BASKETS (make-units 1 1 1)))

;;
;; Exercise 10
;;
;; Rewrite "credit-limit" to run in O(n) time.
;;

;;
;; We will "vectorize" the data structures here:
;;
(defn credit-limit [schedule max-credits]
  (defn credit-limit-iter [sched working total]
    (if (empty? sched)
      (vec working)
      (let [class (first sched)]
	(let [credits (get-class-total-units class)]
	  (if (> (+ credits total) max-credits)
	    (vec working)
	    (vec (credit-limit-iter (rest sched) (concat working [class]) (+ credits total))))))))
  (credit-limit-iter schedule '() 0))

;;
;; Run some unit tests:
;;
(def s2 (empty-schedule))
(def s2 (add-class calculus-1 s1))
(def s2 (add-class algebra s1))
(def s2 (add-class diff-eqs s1))

(total-scheduled-units s2)
;; ==> 30

(credit-limit s2 11)
;; ==> []
(credit-limit s2 12)
;; ==> [{:number CALC-101, :units {:C 4, :L 4, :H 4}}]
(credit-limit s2 20)
;; ==> [{:number CALC-101, :units {:C 4, :L 4, :H 4}}]
(credit-limit s2 21)
;; ==> [{:number CALC-101, :units {:C 4, :L 4, :H 4}} {:number ALGB-152, :units {:C 3, :L 3, :H 3}}]
(credit-limit s2 29)
;; ==> [{:number CALC-101, :units {:C 4, :L 4, :H 4}} {:number ALGB-152, :units {:C 3, :L 3, :H 3}}]
(credit-limit s2 30)
;; ==> [{:number CALC-101, :units {:C 4, :L 4, :H 4}} {:number ALGB-152, :units {:C 3, :L 3, :H 3}} {:number DIFF-201, :units {:C 3, :L 3, :H 3}}]